### PR TITLE
Figure select function selector

### LINKED ISSUE #2629: select traces doesn't do partial matches on `meta`
Today, the output of 

```python
import plotly.graph_objects as go

fig = go.Figure(go.Scatter(meta=dict(a=1,b=2)))

print(len([f for f in fig.select_traces(selector=dict(meta=dict(a=1,b=2)))]))
print(len([f for f in fig.select_traces(selector=dict(meta=dict(a=1)))]))
```

is `1, 0`, meaning that we can't use `selector` to do partial matches on `meta`.

It would be really convenient if PX could tag all of its traces with some structured data in `meta` so that users could target these traces with updates, but for that to be useful we would need to do partial matches.

@jonmmease any major reason not to do this?

### PR DESCRIPTION
**Pull request by nicholas-esterer**
Closes #2629 and more.
Any function calling `BaseFigure._selector_matches` can now pass a function for the `selector` argument. This function is called on each object and if it returns True, then that object is included in the selection.

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Superseded by https://github.com/plotly/plotly.py/pull/2844 I imagine? Can close?

**Comment by nicolaskruchten** [comment]
I meant: supersedes https://github.com/plotly/plotly.py/pull/2836 therefore we can close that one!

File: packages/python/plotly/plotly/basedatatypes.py [hunk 1]
```diff
@@ -813,24 +813,33 @@ def _perform_select_traces(self, filter_by_subplot, grid_subplot_refs, selector)
...
+                if obj_val != selector_val:
+                    return False
+            return True
+        # If selector is a function, call it with the obj as the argument
+        elif type(selector) == type(lambda x: True):
+            return selector(obj)
+        else:
+            raise TypeError(
+                "selector must be dict or a function "
+                "accepting a trace returning a boolean."
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/basedatatypes.py:841]: this is now more than just "a trace" right? maybe "a graph object" ?

**Comment by nicolaskruchten** [comment]
Nicely done!

💃 when you address my comment about `trace` above and you add a changelog entry :)

File: packages/python/plotly/plotly/basedatatypes.py (hunk 1 above)
Comment (nicholas-esterer) [reply packages/python/plotly/plotly/basedatatypes.py:841]: ah yes, copy paste error! good catch :+1:

**Comment by nicolaskruchten** [comment]
Looks great, please merge!

**Pull request merged**